From 239f27377b27690c6c6a10be1d656f287fd73dd6 Mon Sep 17 00:00:00 2001
From: Paul Barker <paul@betafive.co.uk>
Date: Mon, 26 Nov 2018 18:40:08 +0000
Subject: [PATCH] Set DefaultPixelFormat

---
 targetfs/ti335x/etc/powervr.ini | 1 +
 1 file changed, 1 insertion(+)

diff --git a/targetfs/ti335x/etc/powervr.ini b/targetfs/ti335x/etc/powervr.ini
index 29e36ab..f31712a 100644
--- a/targetfs/ti335x/etc/powervr.ini
+++ b/targetfs/ti335x/etc/powervr.ini
@@ -1,2 +1,3 @@
 [default]
 #WindowSystem=libpvrws_WAYLAND.so
+DefaultPixelFormat=RGB565
-- 
2.7.4

